logstash-core/api/lib/app/service.rb

```diff
@@ -0,0 +1,31 @@
+# encoding: utf-8
+require "logstash/instrument/collector"
+
+class LogStash::Api::Service
+
+  include Singleton
+
+  def initialize
+    @snapshot_rotation_mutex = Mutex.new
+    @snapshot = nil
+
+    LogStash::Instrument::Collector.instance.add_observer(self)
+  end
+
+  def stop
+    LogStash::Instrument::Collector.instance.delete_observer(self)
+  end
+
+  def update(time, snapshot)
+    if snapshot_rotation_mutex.try_lock
+      @snapshot = snapshot
+      @snapshot_rotation_mutex.unlock
+    end
+  end
```

Comment:
Thanks for bringing this to me @purbon and sorry for taking more time to review it, Sound like I will revisit thread safety :D

Lets use the Concurrent::AromicReference http://ruby-concurrency.github.io/concurrent-ruby/Concurrent/MutexAtomic.html for now.